### ISSUE TITLE
(Feature request) Boundaries for y values in solve_ivp

### ISSUE THREAD

**Issue by HPLegion** [scipy/scipy#9382]
Hi guys and gals,

first of all thanks for the great work in general - makes my life a lot easier.

I have recently been playing around a lot with the solve_ivp function and the related solvers.

When using the stiff solvers (BDF, Radau), I have a lot of grief with the right hand side of my ODE system being called with "unphysical values", which results in exceptions and / or long run times. If I undertsand this correctly, this is done by the solver to approximate the jacobian. Hence I thought it could be useful if one could define lower and upper bounds on the values of *t* and *y* to tell the solver in which domain the system of equations is well defined and can be called

Looking forward to your opinion,
cheers
Hannes

**Comment by HPLegion** [comment]
Hi Matthew,

thanks for your input. Manipulating the numerical routines to suit my needs sounds like an interesting approach. Unfortunately I currently need to focus on other things, but when I get around to take a look at this again I will keep this thread updated.

Cheers

**Comment by r-menezes** [comment]
Hello there,
I am also using solve_ivp in problems for which it does not make much sense to talk about negative values of `y` . Not sure if it is the same case as @HPLegion , but I guess this problem could be partially solved if it was possible to change the values of `y` in `handle_events`, this way when any variable gets very close to 0, say `y ~ 1e-12`, it would be set to 0, diminishing the change of underflows and keeping the system in some sane region.
Is this a good idea ?

I could help implementing this, with some guidance.

**Comment by drhagen** [comment]
I think this would be a good feature, and I would love to see it in SciPy. However, I am not qualified to implement it. Methods exist, but they are more complex than that proposed by @r-menezes. From the people I have talked to about this, it is not as simple as throwing a `y[x < 0] = 0` as calling it a day. The tweaking of the state has to be done in a way that respects the assumptions of the ODE method or you will get crashes or wrong answers.

**Comment by r-menezes** [comment]
@drhagen , I see...  presumably these assumptions are also different for different methods?
as a workaround, I redefined my diff. equation so that it would return 0 for variables with values below some threshold.
```
def function(time, x_values):
    x_values[x_values < 1e-10] = 0.
    return dxdt(time, x_values)
```
not clean, but its a user tweak that prevents the negative values

**Comment by drhagen** [comment]
> presumably these assumptions are also different for different methods?

I don't know enough about the details of these algorithms, but I don't see why the assumptions could not be different.

**Comment by HPLegion** [comment]
Hi,

> as a workaround, I redefined my diff. equation so that it would return 0 for variables with values below some threshold.
> 
> ```
> def function(time, x_values):
>     x_values[x_values < 1e-10] = 0.
>     return dxdt(time, x_values)



I remember doing something similar and I had mixed results (it probably depends a lot on the ODE itself and just how nonlinear it is). What I found dangerous about this is, that sometimes it would solve the problem but the solution was nonsense (and that was not always immediately obvious). However it has been some time since I dealt with this, and I don't really feel qualified to say too much about this. Consider it more of a friendly advice to be cautious. 

In my case, iirc, I eventually went down a similar path, but instead of steamrolling negative values, I went through all the terms and functions that made up the RHS of my ODEs and adjusted their return values for negative inputs in a way that seemed most sound in terms of the actual physical interpretation of these quantities.

Good luck with your problem and I am intrigued to see what comes out of this. I imagine that this must be a fairly common challenge, so there would probably be some benefit to solving it, but I also acknowledge that the theory behind the solver's is not exactly trivial.

Cheers

**Comment by laurent90git** [comment]
Hi,
I think this is related to #12766, where it was suggested to implement a way to enforce positivity of the variables. I have implemented this for explicit methods and will soon propose a merge with my changes.

Regarding implicit solvers, a limitation of the variables has to be done in the Newton loop (solve_collocation_system for Radau, solve_bdf_system for BDF). Each time a correction dy to the state vector y is computed, the new state vector is norammaly computed as:
y_new = y + dy
Here, we would simply need to allow for a limiter function to be provided and call it in the following way:
y_new = limiter(y+dy)

I can implement this if someone still wants it.

Now this should only be used if we do not counteract the natural behaviour of the ODE, i.e. if the true solution wants to go past the bounds we have set for some variables, then the Newton loop will never converge, and I don't think there's a generic remedy for that. I think in this case the user may do two things:
- redefine the ODE function so that the time derivatives of the the limited variables go smoothly to 0 as these variables approach their bounds
- use an event to detect when a variable reaches its bounds, and then proceed with a corrected ODE function.

Regards

**Comment by HPLegion** [comment]
Hej :)

Thanks for the cross post! 
Indeed my scenario (back then and still occasionally haunting me) was exactly about strictly positive inputs, so it's nice to see there is something happening on that front, I think it is a pretty common constraint.

Regarding the idea to 

> redefine the ODE function so that the time derivatives of the limited variables go smoothly to 0 as these variables approach their bounds

That is the approach I am currently using (with mixed results)

**Comment by laurent90git** [comment]
Hi @HPLegion 

I have tried Radau and BDF on my test problem and they seem to handle it quite well without variable limitation. Do you have like a toy model that reproduce your issue with implicit solvers ?

**Comment by HPLegion** [comment]
Yes Radau generally seems to be extremely stable, have not played around with BDF much, since Radau usually does the trick.

No unfortunately I do not have a simple toy example. It's a pretty messy system of coupled equations made up of numerous (physics model) terms.
